File: frontend/src/actions/suggest-widget.ts
"use server";

import OpenAI from "openai";
import { z } from "zod";

import { Identification } from "@/stores/identification-store";
import { Widget } from "@/stores/widget-store";
import { getUser } from "@/utils/supabase/server";

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY!,
});

const OPENAI_MODEL = "gpt-4o";

export interface SuggestWidgetColumn {
  fieldName: string;
  identification: Identification;
  sampleValues: string[];
}

const widgetSuggestionSchema = z.object({
  name: z.string(),
  description: z.string(),
  vegaLiteSpec: z.record(z.any()),
});

type WidgetSuggestion = z.infer<typeof widgetSuggestionSchema>;

export async function suggestWidget(
  columns: SuggestWidgetColumn[],
  existingWidgets: Widget[],
  dataSize: number
): Promise<WidgetSuggestion> {
  const { user } = await getUser();
  if (!user) throw new Error("Not authenticated");

  if (columns.length > 30) {
    throw new Error("Too many columns. Please limit to 30 columns.");
  }

  const prompt = `Given the following dataset columns with their types and sample
    values, suggest a meaningful Vega-Lite visualization specification.

    # Rules

    1. Be creative to provider the most interesting, useful, data-rich, concise, and
    intuitive visualization. It's OK to provide a targeted visualization of a particular
    aspect of the data -- assume that multiple visualizations will be created. For example,
    a histogram of a particular column is OK.

    2. Do not suggest visualizations that are equivalent or very similar to the
    existing ones listed in the Existing Visualizations section. Choose a different
    perspective or relationship to visualize.

    3. The field names in the Vega-Lite spec must match exactly the fieldNames in the
    Columns section.

    4. The vegaLiteSpec should be a valid Vega-Lite specification in JSON format.

    5. Each column will have ${dataSize} values. Be sure that the visualization
    can render in a reasonable amount of time for this data size, and that the visual
    elements will not overlap or becomes unreadable.

    6. The default width of the visualization will be 500px and the height
    will be 300px. Be sure that the visualization is not too complex for this
    size. For example, do not try to include more than ~ 40 labels on the x-axis
    or y-axis.

    7. The response should be a valid JSON object. It should have the format:

    {
      "name": "...",
      "description": "...",
      "vegaLiteSpec": { ... }
    }

    IT IS VERY IMPORTANT THAT THE RESPONSE IS A ONLY VALID JSON OBJECT.

    # Columns

    ${JSON.stringify(columns, null, 2)}

    # Existing Visualizations

    ${JSON.stringify(
      existingWidgets.map((w) => ({
        name: w.name,
        description: w.description,
      })),
      null,
      2
    )}

    `;

  console.log(prompt);

  let response: string;
  try {
    const completion = await openai.chat.completions.create({
      messages: [{ role: "user", content: prompt }],
      model: OPENAI_MODEL,
      response_format: { type: "json_object" }, // does not work with o1-preview
    });
    const res = completion.choices[0]?.message?.content;
    if (!res) throw new Error("No response from OpenAI");
    response = res;
  } catch (error) {
    console.error("Error suggesting widgets:", error);
    throw new Error("Failed to generate visualization suggestions");
  }

  let parsed: any;
  try {
    parsed = JSON.parse(response);
  } catch (error) {
    console.error("Error parsing response:", error);
    throw new Error("Invalid response format from OpenAI");
  }

  const parseResult = widgetSuggestionSchema.safeParse(parsed);
  if (!parseResult.success) {
    console.error("Invalid response format:", parseResult.error);
    throw new Error("Invalid response format from OpenAI");
  }

  const suggestion = parseResult.data;

  console.log(suggestion);

  return suggestion;
}

File: backend/backend/routers/suggest_widget.py
import json
from typing import Any, List, Literal

from fastapi import APIRouter, Depends, HTTPException
from sqlmodel import SQLModel

from backend import auth
from backend.suggest import create_llm, inference_llm_config

router = APIRouter(
    dependencies=[Depends(auth.get_user_id)],
)

# types (not shared with other routers)

WidgetEngine = Literal["vega-lite", "observable-plot"]


class Identification(SQLModel):
    type: str
    description: str
    suggestedActions: List[str] | None = None
    id: str | None = None
    name: str | None = None
    kind: str | None = None
    minValue: float | None = None
    maxValue: float | None = None
    logScale: bool | None = None


class SuggestWidgetColumn(SQLModel):
    fieldName: str
    identification: Identification
    sampleValues: List[str]


class WidgetSuggestion(SQLModel):
    name: str
    description: str
    engine: WidgetEngine
    vegaLiteSpec: Any | None = None
    observablePlotCode: str | None = None


class SuggestWidgetArgs(SQLModel):
    engine: WidgetEngine
    columns: List[SuggestWidgetColumn]
    existingWidgets: List[WidgetSuggestion]
    dataSize: int


vega_lite_prompt = (
    lambda data_size: f"""
Given the following dataset columns with their types and sample values,
suggest a meaningful Vega-Lite visualization specification.

# Rules

1. Be creative to provide one interesting, useful, concise, and intuitive
   visualization.

2. It's OK to provide a targeted visualization of a particular aspect of the
   data -- assume that multiple visualizations will be created. For example,
   a histogram of a particular column is a good idea if it's interesting.

3. Do not suggest a visualization that is equivalent to one of the existing
   ones listed in the Existing Visualizations section.

4. The field names in the Vega-Lite spec must match exactly the fieldNames in
   the Columns section.

5. The vegaLiteSpec should be a valid Vega-Lite specification in JSON format.

6. Each column will have {data_size} values. Be sure that the visualization
   can render in a reasonable amount of time for this data size, and that the
   visual elements will not overlap or becomes unreadable.  For example, do
   not include more than ~ 40 labels on the x-axis or y-axis or in the
   legend. Do not include more than ~ 400 marks. If the legend is going to be
   too large, consider hiding it and utilizing tooltips.

7. The final visualization will be rendered in a 385px width and 300px height
   container.

8. The response should be a valid JSON object. It should have the format:

{{
  "name": "...",
  "description": "...",
  "vegaLiteSpec": {{ ... }}
}}

IT IS VERY IMPORTANT THAT THE RESPONSE IS A ONLY VALID JSON OBJECT.
"""
)

observable_plot_prompt = (
    lambda data_size: f"""
Given the following dataset columns with their types and sample values, suggest
a meaningful Observable Plot visualization code.

# Rules

1. Be creative to provide one interesting, useful, concise, and intuitive
   visualization.

2. It's OK to provide a targeted visualization of a particular aspect of the
   data -- assume that multiple visualizations will be created. For example, a
   histogram of a particular column is a good idea if it's interesting.

3. Do not suggest a visualization that is equivalent to one of the existing ones
   listed in the Existing Visualizations section.

4. The field names in the Observable Plot code must match exactly the fieldNames
   in the Columns section.

5. The observablePlotCode should be valid JavaScript that creates an Observable
   Plot and appends it to the DOM element with id "root". The code must use the
   global `Plot` object which is already defined. Data is already defined as
   `data` in the global scope. No imports are needed.

6. Each column will have {data_size} values. Be sure that the visualization can
   render in a reasonable amount of time for this data size, and that the visual
   elements will not overlap or becomes unreadable. For example, do not include
   more than ~ 40 labels on the x-axis or y-axis or in the legend. Do not
   include more than ~ 400 marks. If the legend is going to be too large,
   consider hiding it and utilizing tooltips.

7. The final visualization will be rendered in a 385px width and 300px height
   container. The code should append the plot to the DOM element with id "root".

8. The response should be a valid JSON object. It should have the format:

{{
  "name": "...", "description": "...", "observablePlotCode": "..."
}}

IT IS VERY IMPORTANT THAT THE RESPONSE IS A ONLY VALID JSON OBJECT.
"""
)


@router.post("/suggest-widget")
async def suggest_widget(
    args: SuggestWidgetArgs,
) -> WidgetSuggestion:
    """
    Generate a widget suggestion based on the provided columns and existing widgets.
    """

    columns = args.columns
    existing_widgets = args.existingWidgets
    data_size = args.dataSize
    engine = args.engine

    if len(columns) > 30:
        raise HTTPException(status_code=400, detail="Too many columns. Please limit to 30 columns.")

    prompt = (
        vega_lite_prompt(data_size) if engine == "vega-lite" else observable_plot_prompt(data_size)
    )

    prompt += f"""
# Columns

{json.dumps(columns, default=lambda x: x.dict(), indent=2)}

# Existing Visualizations

{json.dumps([{"name": w.name, "description": w.description} for w in existing_widgets], indent=2)}
"""

    print(
        f"🤖 Step 1: Querying {inference_llm_config.model_name} for visualization suggestion..."
        + (
            f" (reasoning_effort: {inference_llm_config.reasoning_effort})"
            if inference_llm_config.reasoning_effort
            else ""
        )
    )

    try:
        # Step 1: Get initial suggestion
        llm = create_llm(inference_llm_config)
        initial_response = await llm.ainvoke(
            prompt,
            response_format=(
                {"type": "json_object"} if inference_llm_config.mode == "structured" else None
            ),
        )
        initial_suggestion = str(initial_response.content)

        print("✅ Initial LLM response received:", initial_suggestion)

        # Parse the response
        parsed = json.loads(initial_suggestion)
        parsed["engine"] = engine

        if engine == "vega-lite" and "vegaLiteSpec" not in parsed:
            raise HTTPException(
                status_code=400,
                detail="Vega-Lite spec not found in the response. Please try again.",
            )

        if engine == "observable-plot" and "observablePlotCode" not in parsed:
            raise HTTPException(
                status_code=400,
                detail="Observable Plot code not found in the response. Please try again.",
            )

        # Validate the response structure
        suggestion = WidgetSuggestion(**parsed)
        return suggestion

    except Exception as error:
        print("❌ Error generating visualization suggestion:", error)
        raise HTTPException(
            status_code=500,
            detail=f"Failed to generate visualization suggestion with {inference_llm_config.model_name}",
        )
